### DISCUSSION TITLE
-Werror causing failures to parse

### DISCUSSION THREAD

**Discussion by mellery451** [clangd/clangd#1919] [Q&A]
context: huge project, lots of warnings enabled for some code, but code is far from warning clean. Many common headers will generate a fair number of warnings, although in many cases we use -isystem for those for just this reason.

observation: if I have -Werror in my compile command for some TU, clang will typically give up with a "too many errors" message and I won't get results for that TU. If I remove this option, clangd completes fine and there are relatively few warnings or errors actually showing up. That is strange in itself. If warnings being converted to error was hitting some error limit, I would still expect those warnings to show-up just with W instead of E. 

a repro is going to be hard given the complexity of this code base, but I'm wondering if anyone else has seen unpleasant failures when -Werror is enabled. I'm going to just remove this flag for now, but I would very much like warnings to be *prominent* during development.

**Comment by HighCommander4** [comment]
You can use the [-ferror-limit](https://clang.llvm.org/docs/UsersManual.html#cmdoption-ferror-limit) flag to increase the number of errors clang will process before giving up with "too many errors".

> If warnings being converted to error was hitting some error limit, I would still expect those warnings to show-up just with W instead of E.

That doesn't seem to be the behaviour of the clang frontend, as seen in this simple example:

test.cpp:
```c++
int main() {
  int unused1;
  int unused2;
  int unused3;
  return 0;
}
```

```
$ clang++ -c -Wunused -Werror -ferror-limit=2 test.cpp
test.cpp:2:7: error: unused variable 'unused1' [-Werror,-Wunused-variable]
    2 |   int unused1;
      |       ^~~~~~~
test.cpp:3:7: error: unused variable 'unused2' [-Werror,-Wunused-variable]
    3 |   int unused2;
      |       ^~~~~~~
fatal error: too many errors emitted, stopping now [-ferror-limit=]
3 errors generated.
```

Clangd just inherits the frontend's behaviour here.

If you'd like to suggest a change to the frontend behaviour, please feel free to file an issue at https://github.com/llvm/llvm-project/issues/.

**Reply by mellery451** [reply]
sorry - I probably wasn't clear in that statement. What I meant was: if I REMOVE -Werror option (but keep the warnings flags), I would expect to see my 20+ warnings that were previously fatal and causing clangd to quit. But that's not the case. I remove -Werror and I then have only one or two warnings from my TU and all is good. 

I suspect something is going on with respect to included headers that generate lots of warnings and may or may not be included with -isystem, but I've tried reproing with your simple case above in a header and cpp and nothing is standing out to me. I'm going to chalk this up to our very messy include/isystem compilation environment (thanks bazel!) and maybe some other mysterious ordering of <> includes. I might also go ahead and try -ferror-limit to see what kinds of results that gives me. thanks.

**Reply by HighCommander4** [reply]
Ah, my bad, I misunderstood. Yeah, that seems strange. I don't have a theory about it but you're able to come up with a repro you can share I'd be curious to look at it.
